File: c++/SingletonPattern/SafeDoubleLockSingleton.cpp
#include "stdafx.h"
#include "SafeDoubleLockSingleton.h"

mutex m_mutex;
SafeDoubleLockSingleton *SafeDoubleLockSingleton::instance = NULL; //静态成员初始化

SafeDoubleLockSingleton::SafeDoubleLockSingleton()
{

}

SafeDoubleLockSingleton *SafeDoubleLockSingleton::GetInstance()
{
	if(instance==NULL)
	{
		m_mutex.lock();
		if(instance==NULL)
		{
			instance = new SafeDoubleLockSingleton();
		}
		m_mutex.unlock();
	}
	return instance;
}

void SafeDoubleLockSingleton::Print()
{
	cout<<"双重锁定,线程安全的单例模式."<<endl;
}

File: csharp/SingletonPattern/SafeDoubleLockSingleton.cs

using System;
namespace SingletonPattern
{
    /// <summary>
    /// 双重锁定，线程安全的单例模式
    /// </summary>
    public sealed class SafeDoubleLockSingleton
    {
        private static SafeDoubleLockSingleton instance;
        private static readonly object syncFlag = new object();

        /// <summary>
        /// 私有的构造函数
        /// </summary>
        private SafeDoubleLockSingleton()
        {

        }

        public static SafeDoubleLockSingleton GetInstance()
        {
            if (instance == null)
            {
                lock (syncFlag)
                {
                    if (instance == null)
                    {
                        instance = new SafeDoubleLockSingleton();
                    }
                }
            }
            return instance;
        }

        public void Print()
        {
            Console.WriteLine("双重锁定,线程安全的单例模式.");
        }
    }
}
